File: languages/csharp/AppEncryption/AppEncryption.IntegrationTests/Multithreaded/MultiFactoryThreadedTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using GoDaddy.Asherah.AppEncryption.IntegrationTests.Utils;
using GoDaddy.Asherah.Logging;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json.Linq;
using Xunit;
using static GoDaddy.Asherah.AppEncryption.IntegrationTests.TestHelpers.Constants;

namespace GoDaddy.Asherah.AppEncryption.IntegrationTests.Multithreaded
{
    [Collection("Configuration collection")]
    public class MultiFactoryThreadedTest
    {
        private static readonly ILogger Logger = LogManager.CreateLogger<MultiFactoryThreadedTest>();
        private readonly ConfigFixture configFixture;

        public MultiFactoryThreadedTest(ConfigFixture configFixture)
        {
            this.configFixture = configFixture;
        }

        // Create multiple sessions from multiple factories
        // to encrypt and decrypt data for multiple partitions.
        // Ensure keys get created properly, DRR's are created as expected,
        // and (future) collect timing information to ensure no thread is
        // starved more than others since there will be some locks in play.
        [Fact]
        public void MultiThreadedMultiFactoryUniquePartitionsEncryptDecrypt()
        {
            // Get the current settings and try to force minWorkers
            ThreadPool.GetMinThreads(out _, out var currentMinIOC);
            Assert.True(ThreadPool.SetMinThreads(NumThreads, currentMinIOC));

            long completedTasks = 0;

            Parallel.ForEach(Enumerable.Range(0, NumRequests), i =>
            {
                try
                {
                    RunPartitionTest(NumIterations, $"request-{i}", PayloadSizeBytes);
                    Interlocked.Increment(ref completedTasks);
                }
                catch (ThreadInterruptedException e)
                {
                    Logger.LogError(e, "Unexpected error during call: ");
                    throw;
                }
            });

            // Wait for all threads to complete
            Assert.Equal(NumRequests, completedTasks);
        }

        // Using the same partition id, create multiple sessions
        // from multiple factories to encrypt and decrypt data.
        // Ensure keys get created properly, DRR's are created as expected,
        // and (future) collect timing information to ensure no thread is
        // starved more than others since there will be some locks in play.
        [Fact]
        public void MultiThreadedMultiFactorySamePartitionEncryptDecrypt()
        {
            // Get the current settings and try to force minWorkers
            ThreadPool.GetMinThreads(out _, out var currentMinIOC);
            Assert.True(ThreadPool.SetMinThreads(NumThreads, currentMinIOC));

            long completedTasks = 0;

            Parallel.ForEach(Enumerable.Range(0, NumRequests), i =>
            {
                try
                {
                    RunPartitionTest(NumIterations, DefaultPartitionId, PayloadSizeBytes);
                    Interlocked.Increment(ref completedTasks);
                }
                catch (ThreadInterruptedException e)
                {
                    Logger.LogError(e, "Unexpected error during call: ");
                    throw;
                }
            });

            // Wait for all threads to complete
            Assert.Equal(NumRequests, completedTasks);
        }

        private void RunPartitionTest(int testIterations, string partitionId, int payloadSizeBytesBase)
        {
            try
            {
                using (SessionFactory sessionFactory =
                    SessionFactoryGenerator.CreateDefaultSessionFactory(
                        configFixture.KeyManagementService,
                        configFixture.Metastore))
                {
                    using (Session<JObject, byte[]> session = sessionFactory.GetSessionJson(partitionId))
                    {
                        Dictionary<string, byte[]> dataStore = new Dictionary<string, byte[]>();

                        string partitionPart = $"partition-{partitionId}-";

                        for (int i = 0; i < testIterations; i++)
                        {
                            // Note the size will be slightly larger since we're adding extra unique meta
                            JObject jObject = PayloadGenerator.CreateRandomJsonPayload(payloadSizeBytesBase);
                            string keyPart = $"iteration-{i}";
                            jObject["payload"] = partitionPart + keyPart;

                            dataStore.Add(keyPart, session.Encrypt(jObject));
                        }

                        foreach (KeyValuePair<string, byte[]> keyValuePair in dataStore)
                        {
                            JObject decryptedObject = session.Decrypt(keyValuePair.Value);
                            Assert.Equal(partitionPart + keyValuePair.Key, decryptedObject["payload"].ToObject<string>());
                        }
                    }
                }
            }
            catch (Exception e)
            {
                Logger.LogError(e, "Unexpected error during call");
                throw;
            }
        }
    }
}

File: tests/java/test-app/src/main/java/com/godaddy/asherah/testapp/multithreaded/MultiFactoryThreadedTest.java
package com.godaddy.asherah.testapp.multithreaded;

import com.godaddy.asherah.appencryption.AppEncryption;
import com.godaddy.asherah.appencryption.AppEncryptionSessionFactory;
import com.godaddy.asherah.testapp.ConfigurationParameterResolver;
import com.godaddy.asherah.testapp.ConfigurationParameterResolver.ConfigurationParameter;
import com.godaddy.asherah.testapp.utils.PayloadGenerator;
import com.godaddy.asherah.testapp.utils.SessionFactoryGenerator;

import org.json.JSONObject;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.LongAdder;
import java.util.stream.IntStream;

import static com.godaddy.asherah.testapp.testhelpers.Constants.*;
import static org.junit.jupiter.api.Assertions.*;

@ExtendWith(ConfigurationParameterResolver.class)
class MultiFactoryThreadedTest {
  private static final Logger LOG = LoggerFactory.getLogger(MultiFactoryThreadedTest.class);

  private void runPartitionTest(final int testIterations, final String partitionId, final int payloadSizeBytesBase) {
    try (AppEncryptionSessionFactory factory = SessionFactoryGenerator.createDefaultAppEncryptionSessionFactory()) {
      try (AppEncryption<JSONObject, byte[]> partition = factory.getAppEncryptionJson(partitionId)) {
        Map<String, byte[]> dataStore = new HashMap<>();

        String partitionPart = "partition-" + partitionId + "-";

        for (int i = 0; i < testIterations; i++) {
          // Note the size will be slightly larger since we're adding extra unique meta
          JSONObject jsonObject = PayloadGenerator.createRandomJsonPayload(payloadSizeBytesBase);
          String keyPart = String.format("iteration-%d", i);
          jsonObject.put("payload", partitionPart + keyPart);

          dataStore.put(keyPart, partition.encrypt(jsonObject));
        }

        dataStore.forEach((key, value) -> {
          JSONObject decryptedObject = partition.decrypt(value);
          assertEquals(partitionPart + key, decryptedObject.get("payload"));
        });
      }
    }
    catch (Exception e) {
      LOG.error("unexpected error during call", e);
      throw e;
    }
  }

  /**
   * Create multiple sessions from multiple factories
   * to encrypt and decrypt data for multiple partitions.
   *
   * Ensure keys get created properly, DRR's are created as expected,
   * and (future) collect timing information to ensure no thread is
   * starved more than others since there will be some locks in play.
   */
  @Test
  void multiThreadedMultiFactoryUniquePartitionsEncryptDecrypt(
      @ConfigurationParameter(TEST_PARAM_NUM_ITERATIONS) final int numIterations,
      @ConfigurationParameter(TEST_PARAM_NUM_REQUESTS) final int numRequests,
      @ConfigurationParameter(TEST_PARAM_NUM_THREADS) final int numThreads,
      @ConfigurationParameter(TEST_PARAM_PAYLOAD_SIZE_BYTES) final int payloadSizeBytes,
      @ConfigurationParameter(TEST_PARAM_THREAD_POOL_TIMEOUT_SECONDS) final long threadPoolTimeoutSeconds) {
    ExecutorService pool = Executors.newFixedThreadPool(numThreads);
    LongAdder tasksCompleted = new LongAdder();

    IntStream.range(0, numRequests)
            .parallel()
            .forEach(i -> {
              pool.submit(() -> {
                runPartitionTest(numIterations, String.format("request-%d", i), payloadSizeBytes);
                tasksCompleted.increment();
              });
            });

    try {
      pool.shutdown();
      pool.awaitTermination(threadPoolTimeoutSeconds, TimeUnit.SECONDS);
    }
    catch (InterruptedException e) {
      LOG.error("interrupted exception during thread pool shutdown", e);
    }

    assertEquals(numRequests, tasksCompleted.intValue());
  }

  /**
   * Using the same partition id, create multiple sessions
   * from multiple factories to encrypt and decrypt data.
   *
   * Ensure keys get created properly, DRR's are created as expected,
   * and (future) collect timing information to ensure no thread is
   * starved more than others since there will be some locks in play.
   */
  @Test
  void multiThreadedMultiFactorySamePartitionEncryptDecrypt(
      @ConfigurationParameter(TEST_PARAM_NUM_ITERATIONS) final int numIterations,
      @ConfigurationParameter(TEST_PARAM_NUM_REQUESTS) final int numRequests,
      @ConfigurationParameter(TEST_PARAM_NUM_THREADS) final int numThreads,
      @ConfigurationParameter(TEST_PARAM_PAYLOAD_SIZE_BYTES) final int payloadSizeBytes,
      @ConfigurationParameter(TEST_PARAM_THREAD_POOL_TIMEOUT_SECONDS) final long threadPoolTimeoutSeconds) {
    ExecutorService pool = Executors.newFixedThreadPool(numThreads);
    LongAdder tasksCompleted = new LongAdder();

    IntStream.range(0, numRequests)
        .parallel()
        .forEach(i -> {
          pool.submit(() -> {
            runPartitionTest(numIterations, DEFAULT_PARTITION_ID, payloadSizeBytes);
            tasksCompleted.increment();
          });
        });

    try {
      pool.shutdown();
      pool.awaitTermination(threadPoolTimeoutSeconds, TimeUnit.SECONDS);
    }
    catch (InterruptedException e) {
      LOG.error("interrupted exception during thread pool shutdown", e);
    }

    assertEquals(numRequests, tasksCompleted.intValue());
  }

}
